primitives/xcm/src/location_matcher.rs

```diff
@@ -0,0 +1,186 @@
+		for_tuples!( #(
+			match Tuple::convert(account.clone()) { o @ Some(_) => return o, _ => () }
+		)* );
+		log::trace!(target: "xcm_primitives::convert", "did not match any location to the account: {:?}", &account);
+		None
+	}
+}
+
+/// A matcher for any address that we would like to compare against a received account.
+/// Tipically used to manage self-reserve currency through the pallet-balances address.
+pub struct SingleAddressMatcher<AccountId, const ADDRESS: u64, PalletInstance>(
+	PhantomData<(AccountId, PalletInstance)>,
+);
+
+impl<AccountId, const ADDRESS: u64, PalletInstance> AccountIdToLocationMatcher<AccountId>
+	for SingleAddressMatcher<AccountId, ADDRESS, PalletInstance>
+where
+	AccountId: Parameter + From<H160>,
+	PalletInstance: PalletInfoAccess,
+{
+	fn convert(account: AccountId) -> Option<Location> {
+		if account == H160::from_low_u64_be(ADDRESS).into() {
+			return Some(Location::new(
+				0,
+				[PalletInstance(
+					<PalletInstance as PalletInfoAccess>::index() as u8,
+				)],
+			));
+		}
+		None
+	}
+}
+
+/// Matcher to compare a received account against some possible foreign asset address.
+pub struct MatchThroughEquivalence<AccountId, AssetId, AssetIdInfoGetter, AssetIdToLocationManager>(
+	PhantomData<(
+		AccountId,
+		AssetId,
+		AssetIdInfoGetter,
+		AssetIdToLocationManager,
+	)>,
+);
+
+impl<AccountId, AssetId, AssetIdInfoGetter, AssetIdToLocationManager>
+	MatchThroughEquivalence<AccountId, AssetId, AssetIdInfoGetter, AssetIdToLocationManager>
+where
+	AccountId: Parameter + Into<H160>,
+	AssetId: From<u8> + TryFrom<u16> + TryFrom<u128>,
+{
+	// Helper function that retrieves the asset_id of a foreign asset account.
+	pub fn account_to_asset_id(
+		account: AccountId,
+		asset_id_info: AssetIdInfo,
+	) -> Option<(Vec<u8>, AssetId)> {
+		let h160_account: H160 = account.into();
+		let (prefix_part, id_part) = h160_account
+			.as_fixed_bytes()
+			.split_at(asset_id_info.size_of + 2);
+
+		if prefix_part == asset_id_info.foreign_asset_prefix {
+			let asset_id: AssetId = match asset_id_info.size_of {
+				2 => {
+					let mut data = [0u8; 2];
+					data.copy_from_slice(id_part);
+					u16::from_be_bytes(data).try_into().unwrap_or(0u8.into())
+				}
+				16 => {
+					let mut data = [0u8; 16];
+					data.copy_from_slice(id_part);
+					u128::from_be_bytes(data).try_into().unwrap_or(0u8.into())
+				}
+				_ => return None,
+			};
+			return Some((prefix_part.to_vec(), asset_id));
+		}
+		None
+	}
+}
+
+impl<AccountId, AssetId, AssetIdInfoGetter, AssetIdToLocationManager>
+	AccountIdToLocationMatcher<AccountId>
+	for MatchThroughEquivalence<AccountId, AssetId, AssetIdInfoGetter, AssetIdToLocationManager>
+where
+	AccountId: Parameter + Into<H160>,
+	AssetId: From<u8> + TryFrom<u16> + TryFrom<u128>,
+	AssetIdInfoGetter: GetAssetIdInfo<AssetId>,
+	AssetIdToLocationManager: MaybeEquivalence<Location, AssetId>,
+{
+	fn convert(account: AccountId) -> Option<Location> {
```

Comment:
The implementation for Foreign Asset is far too complex, tou can make it a lot more simple.
Please remove the type AssetIfInfo and it's trait, instead just take a const generic for the prefix length, and pardse the AssetID as type [u8; PrefixLen], then let final project provide an impl of AssetIdToLocation that will handle this static array

Reply:
Actually, after thinking a bit more about it, I think I have a better and cleaner solution than that one (and also the previous one I created). 

The idea would be to leverage the `AccountIdAssetIdConversion` trait (that currently is owned by the `assets_erc20` precompile) and move it to moonkit `xcm-primitives` package. Both Moonbeam and Tanssi runtimes have this trait implemented, which takes care of managing the asset prefix and the overall complex conversion.

With this approach, we just have to pass a generic parameter to the `ForeignAssetMatcher` that implements this `AccountIdAssetIdConversion` trait, and all the projects that would like to make use of this matcher would have to implement this trait somewhere (in the case of Moonbeam and Tanssi, this trait is already implemented in the runtime, so we don't need extra config on those projects).

In my opinion this approach is very nice, as we avoid duplicating a lot of code, in both Moonbeam and Tanssi runtimes, and also inside the new code we are adding for these matchers.

I already pushed a commit with the changes that put this idea in practice. Let me know if you agree with this solution and the changes look good to you, otherwise we can continue discussing the approach you suggested. 

Thanks in advance!

Reply:
Thank @Agusrodri it's a lot simpler now.